### DISCUSSION TITLE
Best practice to make Edge Config data available globally in a client?

### DISCUSSION THREAD

**Discussion by chranderson** [vercel/community#1250] [Help] [answered]
### Summary

What is the best practice or suggested pattern for accessing edge config values globally from a react hook in the client?
The docs mention edge configs can be used in the client, but all of the examples use server side code & middleware, afaik.

I do have a working implementation, but feel like there is a much more efficient and proper way to handle this.
The connection string env var is private and not avail in the client, so I have a three step process, which seems a bit extra imo.

I would like to avoid making a network request to `/feature-flags` each time I need an update

### Additional information

I understand I can access the edge-config via @vercel/edge-config directly in the next pages/api/feature-flags function, but I want to be able to use it in SSR and SSG if needed. Maybe I am overthinking it?

Server Function to access the config:
```ts
// lib/get-feature-flags.ts

import { createClient } from '@vercel/edge-config'

export interface FeatureFlags {
  showCoolThingy: boolean
}

/** get single feature flag by key */
export async function get(key: keyof FeatureFlags) {
  const edgeConfig = createClient(process.env.FEATURE_FLAGS_CONFIG)
  const featureFlag = await edgeConfig.get<FeatureFlags>(key)
  return featureFlag
}

/** get all feature flags */
export async function getAll() {
  const edgeConfig = createClient(process.env.FEATURE_FLAGS_CONFIG)
  const featureFlags = await edgeConfig.getAll<FeatureFlags>()
  return featureFlags
}

```

Next api endpoint return feature flag data
```ts
// pages/api/feature-flags.ts

import { NextRequest, NextResponse } from 'next/server'
import { get, getAll } from 'lib/get-feature-flags'
import type { FeatureFlags } from 'lib/get-feature-flags'

export default async (req: NextRequest) => {
  const { searchParams } = new URL(req.url)
  const flag = searchParams.get('flag') as keyof FeatureFlags

  /** if no flag provided, return all */
  if (!flag) {
    const featureFlags = await getAll()
    return NextResponse.json(featureFlags)
  }

  const featureFlag = await get(flag)
  return NextResponse.json(featureFlag)
}

export const config = {
  runtime: 'experimental-edge'
}

```

Hook to fetch and provide edge config feature flag data
```ts
// hooks/useFeatureFlags.ts

import { useQuery } from '@tanstack/react-query'
import { useCallback } from 'react'
import type { FeatureFlags } from 'lib/get-feature-flags'

export function useFeatureFlags() {
  const { data: featureFlags } = useQuery<FeatureFlags>(
    ['feature_flags'],
    () => fetch('/api/feature-flags').then(res => res.json()),
    {
      refetchInterval: 300000 // 5 min
    }
  )

  const getFeatureFlag = useCallback(
    (flag: keyof FeatureFlags) =>
      !!featureFlags ? featureFlags[flag] : undefined,
    [featureFlags]
  )

  return {
    /** all feature flags */
    featureFlags,
    /** get single feature flag */
    getFeatureFlag
  }
}

```

Then, when I want to check a feature flag, I can use the hook in **any** component or page.
```ts
// components/CoolComponent

import { useFeatureFlags } from 'hooks/useFeatureFlags'
import { CoolThingyComponent } from 'components/CoolThingy'

export function CoolComponent() {

  const { getFeatureFlag } = useFeatureFlags()
  const showCoolThingy = getFeatureFlag('showCoolThingy')

  return (
    <>
       { showCoolThingy ? <CoolThingyComponent /> : null }
   </>
  )
}
```

**Comment by amyegan** [comment]
Good question @chranderson! I don't have the answer myself, but I'm asking the Edge Config team for their advice on this 🙂

**Reply by chranderson** [reply]
thanks!

**Comment by dferber90** [comment, marked as answer]
### Best performance and avoiding layout shift

I'll preface the answer with some general thoughts on how to get the most out of Edge Config when used for feature flags, but I'll also answer your question in the next section in case you have good reasons to ignore the recommendations.

> I understand I can access the edge-config via https://github.com/orgs/vercel/teams/edge-config directly in the next pages/api/feature-flags function, but I want to be able to use it in SSR and SSG if needed. Maybe I am overthinking it?

When using SSR you can read Edge Config in your server-side code using `@vercel/edge-config`. This incurs no extra latency. We actually expect more and more people to use this pattern as React Server Components get more adoption.

When using SSG it can be extremely powerful to create multiple variants of a static page and rewrite requests to the correct variant from middleware. You can use `getStaticProps` and `getStaticPaths` to make this quite okay to work with. You can then read your Edge Config from middleware to determine which variant of your static page to rewrite a request to.

This is extremely powerful as you
- can read edge config from middleware with negligible latency
- avoid any layout shift since you're dealing with static sites only
  - this is a big deal as Google doesn’t like layout shift and will rank your site lower if you have a lot
- do not need to wait for the client to download any js, parse it and make a network request to determine the right variant

### An alternative solution

I'll eventually go to answer your question but as it's my least favorite approach, let me outline another alternative approach here.

In your middleware, you could read your Edge Config using `@vercel/edge-config`. You could then attach a `Set-Cookie` header to the generated response and pass your feature flags along encoded into a cookie.

On the client you'd then be able to read that cookie without making any network requests.

### How to make an Edge Config public

With all that out of the way, if you have a specific reason to make an Edge Config public, this is how to do that.

> **Warning**: This approach will let anyone read the whole contents of your Edge Config, as you will deliberately leak your Edge Config's read-access token to the client.

If you are okay with having the whole content of your Edge Config public you can let anyone read the Edge Config directly from the frontend by publishing your read access token. Otherwise you could create a separate Edge Config in which you exclusively store content for which it is okay to be accessible publicly.

Let's say your `EDGE_CONFIG` environment variable contains this connection string: `https://edge-config.vercel.com/ecfg_xxx/items?token=yyy`.

You'd then create two new environment variables like so:

```
NEXT_PUBLIC_EDGE_CONFIG_ID=ecfg_xxx
NEXT_PUBLIC_EDGE_CONFIG_READ_ACCESS_TOKEN=yyy
```

When doing this you obviously leak your Edge Config's read-access token as part of that Environment Variable. This gives anyone full read access to your Edge Config. However this is fine if you are okay with the whole content of your Edge Config being available publicly, as the token can only be used to read the Edge Config content.

You can then make a request like this to read your Edge Config directly from your frontend:

```ts
fetch(
  `https://edge-config.vercel.com/${process.env.NEXT_PUBLIC_EDGE_CONFIG_ID}/items`,
  {
    headers: {
      Authorization: `Bearer ${process.env.NEXT_PUBLIC_EDGE_CONFIG_READ_ACCESS_TOKEN}`
    }
  }
)
```

This works by sending a request to the Edge Config's endpoint. Note that this will have a higher latency than reading your Edge Config directly from your Edge Functions, Serverless Functions or Middleware. The Edge Config's endpoint is deployed globally, so you will still have quite low latency but it will not be comparable to reading it from your functions.

### Edge Config Endpoint

This describes the endpoint exposed by Edge Configs in more detail. Each Edge Config exposes an endpoint which is used by `@vercel/edge-config`. You can copy the connection string stored in your `EDGE_CONFIG` environment variable and open it in the browser to query it.

The connection string should look something like `https://edge-config.vercel.com/ecfg_xxx/items?token=yyy`.

#### Authorization header
When you open this URL in the browser you'll see it respond with `{ digest, items }`.  This is the endpoint exposed by the Edge Config itself answering this request.

As it's bad practice to send the request via the URL you should send the token via the Authorization header instead:

```bash
curl -H "Authorization: Bearer yyy" https://edge-config.vercel.com/ecfg_xxx
```

This gives you the same response as before.

#### Reading a specific item

The Edge Config API Endpoint also supports reading an individual item directly like so

```bash
curl -H "Authorization: Bearer yyy" https://edge-config.vercel.com/ecfg_xxx/item/someItemName
```

#### Reading multiple items

```bash
curl -H "Authorization: Bearer yyy" https://edge-config.vercel.com/ecfg_xxx/items?key=itemA&key=itemB
```

This returns itemA and itemB.

#### Reading all items

```bash
curl -H "Authorization: Bearer yyy" https://edge-config.vercel.com/ecfg_xxx/items
```

#### Reading the digest

```bash
curl -H "Authorization: Bearer yyy" https://edge-config.vercel.com/ecfg_xxx/digest
```

This returns the hash of the Edge Config's items.

**Reply by chranderson** [reply]
Thank you @dferber90, I appreciate the explanation. It sounds like I should not be using the edge config as I was hoping to.
I'll look into using React Server Components.

**Reply by dferber90** [reply]
You can totally read Edge Config from React Server Components too if that's helpful

_Just be aware that there's currently a bug which prevents reading Edge Config from React Server Components when using the Edge Runtime, but this will be resolved shortly. It already works when using the default runtime._

**Reply by chranderson** [reply]
@dferber90 Would you suggest to avoid using any of these features (edge-config, react server components, apps directory) in prod at the moment?

**Reply by ChristianIvicevic** [reply]
@dferber90 Sorry to hijack this thread but your sidenote that there is bug with RSC made me curious. I am running into an issue where a simple `await getAll()` in a root layout is causing a mysterious `DYNAMIC_SERVER_USAGE` error without any stacktraces or infos in the function log and is pretty much not debuggable. Is this by any chance the bug you are talking about or something different?

**Reply by dferber90** [reply]
@ChristianIvicevic The issue I am referring to leads to `Error: Failed to get 'integrity' property on 'Request'`, which is something we need to fix internally. Yours seems like a different issue, would you mind opening a separate ticket with more info and ideally a reproduction repo?

**Reply by ChristianIvicevic** [reply]
@dferber90 Thanks for clarification. My issue was more convoluted and is unrelated to edge-config as it was an issue when mixing statically generated app dir pages attempting to run some next-auth code which I was able to fix. Most likely just a PEBCAK :)

**Reply by dferber90** [reply]
@chranderson I've found each team/company has a different culture/policy of when they feel comfortable adopting new technology. The product itself is still in beta and the client library we released is also new, so the APIs might change before GA. There are also no SLAs yet. That said, I feel extremely confident in the stability of Edge Config as the underlying technology has been serving hundreds of millions of request in production for quite a while, and we're actively resolving any issues that come up.

I can't speak too much about the app directory, but you'll probably get a similar answer from the team working on that :) Generally we appreciate everyone jumping on these newer technologies early, as it helps surface those bugs that are only discoverable by having lots of people us the tech in lots of different ways.

**Reply by mwawrusch** [reply]
Adding to the question (and answer), how would this best done with **app** directory? E..g should we place the edge config load into the layout.tsx? Maybe someone has a code sample?

**Reply by ChristianIvicevic** [reply]
@mwawrusch In my root layout I just `await getAll()` the config, validate it with zod and then pass it into a **client-side** context.
```tsx
// In a client-side file
'use client';
const FeaturesContext = createContext<Features>({} as Features);
export const useFeatures = () => useContext(FeaturesContext);

type Props = {
  readonly features: Features;
  readonly children: ReactNode;
};

export const Providers = ({ features, children }: Props) => {
  return (
    <FeaturesContext.Provider value={features}>
      {children}
    </FeaturesContext.Provider>
  );
};

// ...
// In the root layout
<Providers features={features}>{children}</Providers>
```

**Reply by mwawrusch** [reply]
Works like a charm. 
Thank you, much appreciated!
